File: snorky/services/cursors.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

from snorky.services.base import RPCService, RPCError, rpc_command
from snorky.types import values, items, MultiDict
import random


class Cursor(object):
    def __init__(self, private_handle, public_handle, document, position,
                 status, owner):
        self.private_handle = private_handle
        self.public_handle = public_handle
        self.document = document
        self.position = position
        self.status = status
        self.owner = owner

    def for_json(self):
        return {
            "publicHandle": self.public_handle,
            "document": self.document,
            "position": self.position,
            "status": self.status,
            "owner": self.owner.identity,
        }


class CursorsService(RPCService):
    def __init__(self, name):
        RPCService.__init__(self, name)

        # (document : obj) -> set<Client>
        self.clients_by_document = MultiDict()
        # (Client) -> set<document : obj>
        self.documents_by_client = MultiDict()
        # (document : obj) -> dict<public_handle : int, Cursor>
        self.cursors_by_document = {}
        # (Client) -> dict<private_handle : int, Cursor>
        self.cursors_by_client = {}

    def check_identity(self, client):
        if client.identity is None:
            raise RPCError("Not authenticated")

    @rpc_command
    def join(self, req, document):
        client = req.client
        self.check_identity(client)

        if self.clients_by_document.in_set(document, client):
            raise RPCError("Already joined")

        self.clients_by_document.add(document, client)
        self.documents_by_client.add(client, document)
        cursors = values(self.cursors_by_document.get(document, {}))
        return {
            "cursors": [
                cursor.for_json()
                for cursor in cursors
            ]
        }

    @rpc_command
    def leave(self, req, document):
        client = req.client

        if not self.clients_by_document.in_set(document, client):
            raise RPCError("Not joined")

        self.do_leave(client, document)

    def check_new_cursor_allowed(self, client):
        # Limit to 16 cursors per client
        if len(self.cursors_by_client.get(client, {})) >= 16:
            raise RPCError("Too many cursors")

    @rpc_command
    def createCursor(self, req, privateHandle, document, position, status):
        client = req.client

        if not self.clients_by_document.in_set(document, client):
            raise RPCError("Not joined")

        if privateHandle in self.cursors_by_client.get(req.client, {}):
            raise RPCError("Reused handle")

        self.check_new_cursor_allowed(client)

        # Pick a random, free handle
        while True:
            handle = random.getrandbits(16)
            if handle not in self.cursors_by_client.get(client, {}):
                break

        cursor = Cursor(privateHandle, handle,
                        document, position, status, client)

        self.cursors_by_client.setdefault(client, {})
        self.cursors_by_client[client][privateHandle] = cursor
        self.cursors_by_document.setdefault(document, {})
        self.cursors_by_document[document][cursor.public_handle] = cursor

        # Publish
        for other_client in self.clients_by_document.get_set(document):
            if other_client is client:
                continue
            self.send_message_to(other_client, {
                "type": "cursorAdded",
                "cursor": cursor.for_json(),
            })

        return handle

    @rpc_command
    def updateCursor(self, req, privateHandle, newData):
        client = req.client

        if privateHandle not in self.cursors_by_client.get(client, {}):
            raise RPCError("No such cursor")

        cursor = self.cursors_by_client[client][privateHandle]

        if "position" in newData:
            cursor.position = newData["position"]
        if "status" in newData:
            cursor.status = newData["status"]

        # Publish
        for other_client in self.clients_by_document.get_set(cursor.document):
            if other_client is client:
                continue
            self.send_message_to(other_client, {
                "type": "cursorUpdated",
                "cursor": cursor.for_json(),
            })

    @rpc_command
    def removeCursor(self, req, privateHandle):
        client = req.client

        if privateHandle not in self.cursors_by_client.get(client, {}):
            raise RPCError("No such cursor")

        self.do_remove_cursor(client, privateHandle)

    def do_remove_cursor(self, client, private_handle):
        cursor = self.cursors_by_client[client][private_handle]

        del self.cursors_by_document[cursor.document][cursor.public_handle]
        if len(self.cursors_by_document[cursor.document]) == 0:
            del self.cursors_by_document[cursor.document]

        del self.cursors_by_client[client][private_handle]
        if len(self.cursors_by_client[client]) == 0:
            del self.cursors_by_client[client]

        # Publish
        for other_client in self.clients_by_document.get_set(cursor.document):
            if other_client is client:
                continue
            self.send_message_to(other_client, {
                "type": "cursorRemoved",
                "cursor": {
                    "publicHandle": cursor.public_handle,
                    "document": cursor.document,
                },
            })

    def do_leave(self, client, document):
        # Get list of cursor handles belonging to the specified user and
        # document
        handles = [private_handle for (private_handle, cursor)
                   in items(self.cursors_by_client.get(client, {}))
                   if cursor.document == document]

        for private_handle in handles:
            self.do_remove_cursor(client, private_handle)

        self.clients_by_document.remove(document, client)
        self.documents_by_client.remove(client, document)

    def client_disconnected(self, client):
        documents = list(self.documents_by_client.get_set(client))

        for document in documents:
            self.do_leave(client, document)

File: snorky/frontend/src/services/cursors.js
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

(function() {
  "use strict";

  var Class = Snorky.Class;
  var _ = Snorky._;

  var Cursor = new Class({});

  var ForeignCursor = new Class(Cursor, {
    constructor: function(data) {
      _.assign(this, data);
    },

    isMine: false
  });

  var OwnCursor = new Class(Cursor, {
    constructor: function(document, data) {
      this.document = document;
      this.removed = false;
      _.assign(this, data);

      this.pendingPromiseCount = 1; // The create request is pending still
    },

    isMine: true,

    update: function(newData) {
      var self = this;
      var service = this.document.service;
      self.pendingPromiseCount += 1;

      _.subset(newData, ["position", "status"], function(key) {
        throw new Error("Can't update '" + key + "' of cursor.");
      });

      // Optimization: skip rpcCall if old data and new data are equal
      if (_.isEqualRight(this, newData)) {
        return new Promise(function(success) { success(); });
      } else {
        _.assign(this, newData);

        return service.rpcCall("updateCursor", {
          privateHandle: this.privateHandle,
          newData: newData
        }).then(function () {
          self.pendingPromiseCount -= 1;
        });
      }
    },

    remove: function() {
      var self = this;
      var service = this.document.service;
      self.pendingPromiseCount += 1;
      self.removed = true;

      delete service.ownCursors[self.privateHandle];
      _.removeOnField("privateHandle", this.document.cursors,
                      this.privateHandle);

      return service.rpcCall("removeCursor", {
        privateHandle: this.privateHandle
      }).then(function () {
        self.pendingPromiseCount -= 1;
      });
    },

    isSynchronized: function() {
      return this.pendingPromiseCount === 0;
    }
  });

  var Document = new Class({
    constructor: function(service, name, cursors) {
      this.service = service;
      this.name = name;

      this.cursors = _.map(cursors, function(cursor) {
        cursor.document = this;
        return new ForeignCursor(cursor);
      }, this);

      this.cursorAdded = new Snorky.Signal();
      this.cursorUpdated = new Snorky.Signal();
      this.cursorRemoved = new Snorky.Signal();
    },

    onCursorAdded: function(cursor) {
      var fatCursor = new ForeignCursor(cursor);
      this.cursors.push(fatCursor);

      this.cursorAdded.dispatch(fatCursor)
    },

    onCursorUpdated: function(cursor) {
      var fatCursor = _.replaceOnField("publicHandle", this.cursors,
                                       cursor);
      this.cursorUpdated.dispatch(fatCursor);
    },

    onCursorRemoved: function(cursor) {
      _.removeOnField("publicHandle", this.cursors, cursor);
      this.cursorRemoved.dispatch(cursor);
    },

    createCursor: function(params) {
      var handle = this.service.getFreePrivateHandle();

      var cursor = new Snorky.Cursors.OwnCursor(this, {
        privateHandle: handle,
        position: params.position,
        status: params.status
      });

      var promise = this.service.rpcCall("createCursor", {
        privateHandle: handle,
        document: this.name,
        position: params.position,
        status: params.status
      }).then(function(publicHandle) {
        cursor.publicHandle = publicHandle;
        cursor.pendingPromiseCount -= 1;
      });

      this.service.ownCursors[handle] = cursor;
      this.cursors.push(cursor);

      return [cursor, promise];
    }
  });

  Snorky.Cursors = new Class(Snorky.RPCService, {
    init: function() {
      Snorky.RPCService.prototype.init.call(this);

      this.notificationReceived.add(this.onNotification, this);

      this.documents = {};
      this.ownCursors = {};
    },

    join: function(params) {
      var self = this;
      return this.rpcCall("join", params)
        .then(function(documentData) {
          var name = params.document;
          var document = new Snorky.Cursors.Document(
            self, name, documentData.cursors);
          self.documents[stableStringify(name)] = document;
          return document;
        });
    },

    onNotification: function(notification) {
      var cursor = notification.cursor;
      var document = this.documents[stableStringify(cursor.document)];
      if (!document) {
        throw Error("Notification for unknown document: " +
                    stableStringify(notification.document));
      }

      if (notification.type == "cursorAdded") {
        document.onCursorAdded(cursor);
      } else if (notification.type == "cursorUpdated") {
        document.onCursorUpdated(cursor);
      } else if (notification.type == "cursorRemoved") {
        document.onCursorRemoved(cursor);
      }
    },

    getFreePrivateHandle: function() {
      if (_.keys(this.ownCursors).length > 100) {
        throw new Error("Too many cursors (thrown at client side)");
      }
      var handle;
      do {
        handle = Math.round(Math.random() * 255);
      } while (handle in this.ownCursors);
      return handle;
    }
  });

  Snorky.Cursors.Document = Document;
  Snorky.Cursors.Cursor = Cursor;
  Snorky.Cursors.ForeignCursor = ForeignCursor;
  Snorky.Cursors.OwnCursor = OwnCursor;

})();


